rust/src/bittorrent_dht/bittorrent_dht.rs

```diff
@@ -0,0 +1,361 @@
+    tx: *mut std::os::raw::c_void, _direction: u8,
+) -> std::os::raw::c_int {
+    let tx = cast_pointer!(tx, BitTorrentDHTTransaction);
+
+    // Transaction is done if we have a request, response, or error since
+    // a new transaction is created for each received packet
+    if tx.request.is_some() || tx.response.is_some() || tx.error.is_some() {
+        return 1;
+    }
+    return 0;
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_bittorrent_dht_state_get_events(
+    tx: *mut std::os::raw::c_void,
+) -> *mut core::AppLayerDecoderEvents {
+    let tx = cast_pointer!(tx, BitTorrentDHTTransaction);
+    return tx.events;
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_bittorrent_dht_state_get_tx_iterator(
+    _ipproto: u8, _alproto: AppProto, state: *mut std::os::raw::c_void, min_tx_id: u64,
+    _max_tx_id: u64, istate: &mut u64,
+) -> applayer::AppLayerGetTxIterTuple {
+    let state = cast_pointer!(state, BitTorrentDHTState);
+    match state.tx_iterator(min_tx_id, istate) {
+        Some((tx, out_tx_id, has_next)) => {
+            let c_tx = tx as *const _ as *mut _;
+            let ires = applayer::AppLayerGetTxIterTuple::with_values(c_tx, out_tx_id, has_next);
+            return ires;
+        }
+        None => {
+            return applayer::AppLayerGetTxIterTuple::not_found();
+        }
+    }
+}
+
+// Parser name as a C style string.
+const PARSER_NAME: &[u8] = b"bittorrent-dht\0";
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_bittorrent_dht_udp_register_parser() {
+    let default_port = CString::new("[1024:65535]").unwrap();
+    let parser = RustParser {
+        name: PARSER_NAME.as_ptr() as *const std::os::raw::c_char,
+        default_port: default_port.as_ptr(),
+        ipproto: IPPROTO_UDP,
+        probe_ts: Some(rs_bittorrent_dht_probing_parser),
+        probe_tc: Some(rs_bittorrent_dht_probing_parser),
+        min_depth: 0,
+        max_depth: 16,
+        state_new: rs_bittorrent_dht_state_new,
+        state_free: rs_bittorrent_dht_state_free,
+        tx_free: rs_bittorrent_dht_state_tx_free,
+        parse_ts: rs_bittorrent_dht_parse,
+        parse_tc: rs_bittorrent_dht_parse,
+        get_tx_count: rs_bittorrent_dht_state_get_tx_count,
+        get_tx: rs_bittorrent_dht_state_get_tx,
+        tx_comp_st_ts: 1,
+        tx_comp_st_tc: 1,
+        tx_get_progress: rs_bittorrent_dht_tx_get_alstate_progress,
+        get_de_state: rs_bittorrent_dht_tx_get_detect_state,
+        set_de_state: rs_bittorrent_dht_tx_set_detect_state,
+        get_events: Some(rs_bittorrent_dht_state_get_events),
+        get_eventinfo: Some(BitTorrentDHTEvent::get_event_info),
+        get_eventinfo_byid: Some(BitTorrentDHTEvent::get_event_info_by_id),
+        localstorage_new: None,
+        localstorage_free: None,
+        get_files: None,
+        get_tx_iterator: Some(rs_bittorrent_dht_state_get_tx_iterator),
+        get_tx_data: rs_bittorrent_dht_get_tx_data,
+        apply_tx_config: None,
+        flags: 0,
```

Comment:
transactions look unidirectional, so `APP_LAYER_PARSER_OPT_UNIDIR_TXS` should be set here